### PR TITLE
Radio test project

### LINKED ISSUE #277: Add Radio Project to Dotbot
Test of the Radio

### PR DESCRIPTION
**Pull request by HappyTakoumi**
fixes #277

### CODE REVIEW

**Comment by aabadie** [comment]
#236 also modified the Makefile and was already merged. You have to rebase:
1. rebase to current main
```
git pull --rebase origin main
```
2. Fix potential conflicts, manually or use a merge tool
```
git mergetool
```
3. Finalize the rebase
```
git rebase --continue
```
4. Push your PR branch
```
git push raphael radio_test_project -f
```

File: projects/03app_radio_test/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Radioapplication
+
+## This application allows test for the radio see at https://github.com/HappyTakoumi/nRF52840DK-radio-Check
+
+### In this repertory you can found:
```
Comment (aabadie) [inline projects/03app_radio_test/README.md:5]:
```suggestion
### In this directory you can found:
```

File: projects/03app_radio_test/README.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+* TX: Transmit packet
+* RX: Receive packet and send them in the uart
```
Comment (aabadie) [inline projects/03app_radio_test/README.md:7]:
```suggestion
* RX: Receive radio packets and forwards them to UART
```

File: projects/03app_radio_test/main_Rx.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    db_radio_set_frequency(8);  // Set the RX frequency to 2408 MHz.
+    db_radio_set_crc_callback(&_radio_callback_crc);
+    db_radio_rx();  // Start receiving packets.
+
+    db_uart_init(DB_UART_INDEX, &db_uart_rx, &db_uart_tx, DB_UART_BAUDRATE, &_uart_callback);
+
+    while (1) {
+    }
+}
+/*************************** End of file ****************************/
```
Comment (aabadie) [inline projects/03app_radio_test/main_Rx.c:72]: This is useless
```suggestion
```

File: projects/03app_radio_test/main_Tx.c [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    uint8_t packet_counter = 0;
+
+    while (1) {
+        db_radio_disable();
+        db_radio_tx(&packet_counter, sizeof(uint8_t));
+        packet_counter++;
+        db_timer_delay_ms(500);
+    }
+}
```
Comment (aabadie) [inline projects/03app_radio_test/main_Tx.c:50]: Missing new line at end of file:
```suggestion
}

```
You can configure your editor to add newline automatically when saving files.

File: projects/03app_radio_test/main_Tx_Idle.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    db_protocol_init();
+    db_radio_init(&_radio_callback, DB_RADIO_BLE_1MBit);
+    db_radio_set_frequency(8);  // Set the RX frequency to 2408 MHz.
+    db_radio_set_power(RADIO_TXPOWER_TXPOWER_0dBm);
+    db_radio_txidle_state();  // Start receiving packets.
+
+    while (1) {
+    }
+}
```
Comment (aabadie) [inline projects/03app_radio_test/main_Tx_Idle.c:46]: Missing new line

File: projects/projects-radio-test.emProject [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+      <file file_name="$(DeviceMemoryMap)">
+        <configuration Name="Common" file_type="Memory Map" />
+      </file>
+      <file file_name="../../nRF/Scripts/nRF_Target.js">
+        <configuration Name="Common" file_type="Reset Script" />
+      </file>
+    </folder>
+    <folder Name="Source Files">
+      <configuration Name="Common" filter="c;cpp;cxx;cc;h;s;asm;inc" />
+      <file file_name="main_Rx.c" />
```
Comment (aabadie) [inline projects/projects-radio-test.emProject:32]: Use lower case file names, maybe also add "radio_test" to the name:
```suggestion
      <file file_name="main_radio_test_rx.c" />
```

File: projects/projects-radio-test.emProject [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      <file file_name="$(DeviceMemoryMap)">
+        <configuration Name="Common" file_type="Memory Map" />
+      </file>
+      <file file_name="../../nRF/Scripts/nRF_Target.js">
+        <configuration Name="Common" file_type="Reset Script" />
+      </file>
+    </folder>
+    <folder Name="Source Files">
+      <configuration Name="Common" filter="c;cpp;cxx;cc;h;s;asm;inc" />
+      <file file_name="main_Tx.c" />
```
Comment (aabadie) [inline projects/projects-radio-test.emProject:71]:
```suggestion
      <file file_name="main_radio_test_tx.c" />
```

File: projects/projects-radio-test.emProject [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      <file file_name="$(DeviceMemoryMap)">
+        <configuration Name="Common" file_type="Memory Map" />
+      </file>
+      <file file_name="../../nRF/Scripts/nRF_Target.js">
+        <configuration Name="Common" file_type="Reset Script" />
+      </file>
+    </folder>
+    <folder Name="Source Files">
+      <configuration Name="Common" filter="c;cpp;cxx;cc;h;s;asm;inc" />
+      <file file_name="main_Tx_Idle.c" />
```
Comment (aabadie) [inline projects/projects-radio-test.emProject:110]:
```suggestion
      <file file_name="main_radio_test_tx_idle.c" />
```

File: projects/projects-radio-test.emProject [hunk 9]
```diff
@@ -0,0 +1,120 @@
+<!DOCTYPE CrossStudio_Project_File>
+<solution Name="projects-radio-test" target="8" version="2">
```
Comment (aabadie) [inline projects/projects-radio-test.emProject:2]:
```suggestion
<solution Name="Radio Test" target="8" version="2">
```

File: projects/projects-radio-test.emProject [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+  <project Name="project_Rx">
```
Comment (aabadie) [inline projects/projects-radio-test.emProject:3]:
```suggestion
  <project Name="RX">
```

File: projects/projects-radio-test.emProject [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    </folder>
+    <folder Name="System Files">
+      <file file_name="$(DeviceCommonVectorsFile)" />
+      <file file_name="$(DeviceVectorsFile)">
+        <configuration Name="Common" file_type="Assembly" />
+      </file>
+      <file file_name="$(SeggerThumbStartup)" />
+    </folder>
+  </project>
+  <project Name="project_Tx">
```
Comment (aabadie) [inline projects/projects-radio-test.emProject:42]:
```suggestion
  <project Name="TX">
```

File: projects/projects-radio-test.emProject [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    </folder>
+    <folder Name="System Files">
+      <file file_name="$(DeviceCommonVectorsFile)" />
+      <file file_name="$(DeviceVectorsFile)">
+        <configuration Name="Common" file_type="Assembly" />
+      </file>
+      <file file_name="$(SeggerThumbStartup)" />
+    </folder>
+  </project>
+  <project Name="project_Tx_Idle">
```
Comment (aabadie) [inline projects/projects-radio-test.emProject:81]:
```suggestion
  <project Name="TX idle">
```

File: projects/03app_radio_test/main_Rx.c [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    db_protocol_init();
+    db_radio_init(&_radio_callback, DB_RADIO_BLE_1MBit);
+    db_radio_set_frequency(8);  // Set the RX frequency to 2408 MHz.
+    db_radio_set_crc_callback(&_radio_callback_crc);
+    db_radio_rx();  // Start receiving packets.
+
+    db_uart_init(DB_UART_INDEX, &db_uart_rx, &db_uart_tx, DB_UART_BAUDRATE, &_uart_callback);
+
+    while (1) {
+    }
```
Comment (aabadie) [inline projects/03app_radio_test/main_Rx.c:70]:
```suggestion
    while (1) {}
```

File: projects/03app_radio_test/main_Rx.c [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+static void _radio_callback_crc(uint8_t *packet, uint8_t length) {
+    memcpy(packet + length, &write_crc, 3);
+    db_uart_write(DB_UART_INDEX, packet, length + 3);
+}
+
+static void _radio_callback(uint8_t *packet, uint8_t length) {
+    uint8_t RSSI = db_radio_rssi();
+    memcpy(packet + length, &RSSI, 1);
+    memcpy(packet + length + 1, &uart_end, 1);
```
Comment (aabadie) [inline projects/03app_radio_test/main_Rx.c:47]:
```suggestion
    packet[length++] = RSSI;
    packet[length++] = uart_end;
```

Note that this is a bit ugly because this code is writing in a buffer that is allocated by the radio driver. So you could have an overflow here. But in your case that is fine because the driver buffer size is 255 and you know you should only receive 1 byte (length = 1)

File: projects/03app_radio_test/main_Tx_Idle.c [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    db_timer_set_periodic_ms(0, 2000, _led1_blink_slow);
+
+    db_protocol_init();
+    db_radio_init(&_radio_callback, DB_RADIO_BLE_1MBit);
+    db_radio_set_frequency(8);  // Set the RX frequency to 2408 MHz.
+    db_radio_set_power(RADIO_TXPOWER_TXPOWER_0dBm);
+    db_radio_txidle_state();  // Start receiving packets.
+
+    while (1) {
+    }
```
Comment (aabadie) [inline projects/03app_radio_test/main_Tx_Idle.c:45]:
```suggestion
    while (1) {}
```

File: projects/03app_radio_test/main_Rx.c [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+#include <stdlib.h>
+#include <string.h>
+// Include BSP headers
+#include "uart.h"
+#include "board.h"
+#include "board_config.h"
+#include "gpio.h"
+#include "protocol.h"
+#include "radio.h"
+#include "timer.h"
```
Comment (aabadie) [inline projects/03app_radio_test/main_Rx.c:22]:
```suggestion
#include "board.h"
#include "board_config.h"
#include "gpio.h"
#include "radio.h"
#include "timer.h"
#include "uart.h"
```

File: projects/03app_radio_test/main_Tx.c [hunk 17]
```diff
@@ -0,0 +1,50 @@
+/**
+ * @file
+ * @defgroup project_Radio    Radio application
```
Comment (aabadie) [inline projects/03app_radio_test/main_Tx.c:3]:
```suggestion
 * @defgroup project_Radio    Radio TX test application
```

File: projects/03app_radio_test/main_Tx_Idle.c [hunk 18]
```diff
@@ -0,0 +1,46 @@
+/**
+ * @file
+ * @defgroup project_Radio    Radio application
```
Comment (aabadie) [inline projects/03app_radio_test/main_Tx_Idle.c:3]:
```suggestion
 * @defgroup project_Radio    Radio TX idle test application
```

**Review by aabadie — COMMENTED** [review]
A first round of review. I haven't checked yet how to fix the CI issue.

File: Makefile [hunk 19]
```diff
@@ -164,4 +164,4 @@ docclean:
 	make -C doc/sphinx clean --no-print-directory
 
 doc:
-	make -C doc/sphinx linkcheck html --no-print-directory SPHINXOPTS="-W --keep-going -n"
+	make -C doc/sphinx linkcheck html --no-print-directory SPHINXOPTS="-W --keep-going -n"
```
Comment (aabadie) [inline Makefile:167]: unrelated change

File: nrf52840dk.emProject [hunk 20]
```diff
@@ -63,5 +63,10 @@
   <import file_name="projects/projects-gateway.emProject" />
   <import file_name="projects/projects-log-dump.emProject" />
   <import file_name="projects/projects-sailbot.emProject" />
+<<<<<<< HEAD
   <import file_name="otap/otap.emProject" />
+=======
+  <import file_name="projects/projects-radio-test.emProject" />
+  <import file_name="testbed/testbed.emProject" />
+>>>>>>> f81dab7 (nrf58420dk: update for project radio_test)
```
Comment (aabadie) [inline nrf52840dk.emProject:71]: There's a merge conflict here
```suggestion
  <import file_name="projects/projects-radio-test.emProject" />
  <import file_name="otap/otap.emProject" />
```

File: projects/03app_radio_test/README.md [hunk 21]
```diff
@@ -0,0 +1,10 @@
+# Radioapplication
```
Comment (aabadie) [inline projects/03app_radio_test/README.md:1]:
```suggestion
# Radio test application
```

File: projects/03app_radio_test/README.md [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+
+## This application allows test for the radio see at https://github.com/HappyTakoumi/nRF52840DK-radio-Check
```
[the remark this answers is not in the document]
Comment (HappyTakoumi) [reply projects/03app_radio_test/README.md:3]: where to I put my python code that go with this ?

[the remark this answers is not in the document]
Comment (aabadie) [reply projects/03app_radio_test/README.md:3]: Good point! I forgot to mention the Python script. You can add it in `dist/scripts/radio_test`

File: bsp/nrf/radio.c [hunk 23]
```diff
@@ -122,9 +123,9 @@ void db_radio_init(radio_cb_t callback, db_radio_ble_mode_t mode) {
                            (RADIO_PCNF1_WHITEEN_Enabled << RADIO_PCNF1_WHITEEN_Pos);  // Enable data whitening feature.
     } else {                                                                          // Long ranges modes (125KBit/500KBit)
 #if defined(NRF5340_XXAA)
-        NRF_RADIO->TXPOWER = (RADIO_TXPOWER_TXPOWER_0dBm << RADIO_TXPOWER_TXPOWER_Pos);  // 0dBm Power output
+        db_radio_set_power(RADIO_TXPOWER_TXPOWER_0dBm);  // 0dBm Power output
 #else
-        NRF_RADIO->TXPOWER = (RADIO_TXPOWER_TXPOWER_Pos8dBm << RADIO_TXPOWER_TXPOWER_Pos);  // 8dBm Power output
+        db_radio_set_power(RADIO_TXPOWER_TXPOWER_0dBm);  // 8dBm Power output
```
Comment (aabadie) [inline bsp/nrf/radio.c:128]:
```suggestion
        db_radio_set_power(RADIO_TXPOWER_TXPOWER_Pos8dBm);  // 8dBm Power output
```

File: bsp/radio.h [hunk 24]
```diff
@@ -67,6 +75,15 @@ void db_radio_set_frequency(uint8_t freq);
  */
 void db_radio_set_channel(uint8_t channel);
 
+/**
+ * @brief Set the power of the radio
+ *
+ * By default 0 dBm
```
Comment (aabadie) [inline bsp/radio.h:81]: There's no default value AFAICT

File: bsp/radio.h [hunk 25]
```diff
@@ -98,6 +115,15 @@ void db_radio_tx(const uint8_t *packet, uint8_t length);
  */
 void db_radio_rx(void);
 
+/**
+ * @brief Block Radio into TXidle state
```
Comment (aabadie) [inline bsp/radio.h:119]:
```suggestion
 * @brief Block Radio into TX idle state
```

File: dist/scripts/radio_test/nRF58420DK_radio_packets_check.py [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+
+print("the first and last packets missing will not be know") 
+result = 'Total Packets Receive :' + str(len(packets))
+print(result) 
+result = 'Missing Packets :' + str(miss)
+print(result) 
+result = 'CRC Packets :' + str(crc)
+print(result) 
+result = 'RSSI dBm :' + str( np.int8((sum(RSSI))/len(RSSI)))
+print(result)
```
Comment (aabadie) [inline dist/scripts/radio_test/nRF58420DK_radio_packets_check.py:76]: Still missing newline.

Use a real IDE like vscode and configure it with newline automatically added on file saved. Notepad++ is too limited.
For Python development, vscode as all the extensions you need to make you dev life better

Comment (aabadie) [inline dist/scripts/radio_test/nRF58420DK_radio_packets_check.py:76]: Use Python f-strings in a single call to `print()`. Python support multi-line strings:
```py
var  = "example"
print(
    f"This is an f-string that prints {example}"
    "This is a regular string"
)
```

File: doc/sphinx/_static/images/python_radio_test_demo.JPG
Comment (aabadie) [inline doc/sphinx/_static/images/python_radio_test_demo.JPG:1]: use lower-case extension (here it's JPG)

File: radio_test/README.md [hunk 27]
```diff
@@ -0,0 +1,23 @@
+# GOAL: Check the radio communication between two nRF52840DK
```
Comment (aabadie) [inline radio_test/README.md:1]: The radio test can work on any nrf52 and nrf5340. It's not limited to nrf52840dk

```suggestion
# GOAL: Check the radio communication between two nRF boards
```

File: radio_test/README.md [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+
+## Segger:
+  * **UnZIP** then open `DotBot-firmware` file and open project called `nRF52840dk` with Segger
+  * Tx: **build** then **load** `projects_radio_test -> project Tx` on the first nRF
+  * Rx: **build** then **load** `projects_radio_test -> project Rx` on the second nRF
+  * Tx_Idle: **build** then **load** `projects_radio_test -> project TxIdle` on the third nRF (you can change the power for your test)
```
Comment (aabadie) [inline radio_test/README.md:7]: I don't think the projects are still called this way

File: radio_test/README.md [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+
+  ![Démo](../doc/sphinx/_static/images/radio_test_setup.jpg)
+    
+ ## Python /Conda:
+  * **download** and **install** Anaconda 
+  * **download** in the python file in **dist/scripts/radio_test**
+  * With Conda: `conda install anaconda::pyserial` (go next if already done)
```
Comment (aabadie) [inline radio_test/README.md:14]: conda is not a standard Python setup. People can use it of course.
It's not meaningful to tell people to install conda, users are supposed to know how to do that. Just give the minimal Python version supported (Python 3.10+ is a safe bet).
To install Python packages give `pip` based command since this is the standard Python package installer.

File: radio_test/README.md [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+  * With **cd** command go to directory where you have downloaded /move the python file (use TAB to help fill the mane of directory and **dir** (on windows) / **ls** (linux) to know the list files on the directory)
+  * With the peripheral manage **find** the COM of your nRF with the RX code
+  * With text file editor **change** the COM  with the write one
+  * **Launch** the script: _python test_radio.py_
```
Comment (aabadie) [inline radio_test/README.md:18]: If you add `#!/usr/bin/env python` (this is called shebang) and the very beginning of the script and give executable permission to the script, you don't have to use "python" and can just call it directly.

```suggestion
  * **Launch** the script: _test_radio.py_
```

File: radio_test/README.md [hunk 31]
```diff
@@ -0,0 +1,23 @@
...
+  * The script will end when you disconnect the nRF with the Rx code and give you the result of the radio communication( CRC, Error, number of packets)
```
Comment (aabadie) [inline radio_test/README.md:19]:
```suggestion
  * The script ends when you disconnect the nRF with the Rx code and gives you the result of the radio communication( CRC, Error, number of packets)
```

File: radio_test/rx/main_radio_test_rx.c [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+#include <string.h>
+// Include BSP headers
+#include "uart.h"
+#include "gpio.h"
+#include "board_config.h"
+#include "radio.h"
+#include "timer.h"
+
+uint8_t write_crc[3] = "CRC";
+uint8_t uart_end     = '\x00';
```
Comment (aabadie) [inline radio_test/rx/main_radio_test_rx.c:23]: (already commented about in my previous review)

```suggestion
static const uint8_t write_crc[3] = "CRC";
static const uint8_t uart_end     = 0x00;
```

File: dist/scripts/radio_test/nRF58420DK_radio_packets_check.py [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+
+
+#check list
+for i in range(len(packets)-1) :
+    
+    if not isinstance(packets[i], int) : # if not int means string = CRCFAIl
+        crc+= 1  
+        
+
+print("the first and last packets missing will not be know")
```
Comment (aabadie) [inline dist/scripts/radio_test/nRF58420DK_radio_packets_check.py:68]: What does that mean?

File: dist/scripts/radio_test/nRF58420DK_radio_packets_check.py [hunk 34]
```diff
@@ -0,0 +1,76 @@
+#test_radio.py
```
Comment (aabadie) [inline dist/scripts/radio_test/nRF58420DK_radio_packets_check.py:1]: Install black and run it on that script:
```
pip install black
black nRF52840DK_radio_packets_check.py
```

Rename the script `radio_test.py`

File: dist/scripts/radio_test/nRF58420DK_radio_packets_check.py [hunk 35]
```diff
@@ -0,0 +1,76 @@
...
+x=['0','1','0','0']
+cnt=0
+crc=0
+error=0
+
+
+myports = [tuple(p) for p in list(serial.tools.list_ports.comports())]
+nRF_rx = serial.Serial(serialport_rx, baud, timeout=1)
+serialport_rx  = [port for port in myports if serialport_rx  in port ][0]
+
```
Comment (aabadie) [inline dist/scripts/radio_test/nRF58420DK_radio_packets_check.py:22]: This script has to be rewritten entirely.

**Comment by aabadie** [comment]
The CI build jobs will pass with the following diff:

```diff
diff --git a/Makefile b/Makefile
index aa6c471..517cc89 100644
--- a/Makefile
+++ b/Makefile
@@ -72,17 +72,20 @@ RADIO_TEST_APPS ?= $(shell find radio_test/ -maxdepth 1 -mindepth 1 -type d | tr
 ifneq (,$(filter dotbot-v1,$(BUILD_TARGET)))
   PROJECTS := $(filter-out 01bsp_qdec 01drv_lis3mdl 01drv_move 03app_dotbot_gateway 03app_dotbot_gateway_lr 03app_sailbot 03app_nrf5340_%,$(PROJECTS))
   ARTIFACT_PROJECTS := 03app_dotbot
+  RADIO_TEST_APPS :=
 endif
 
 ifneq (,$(filter dotbot-v2,$(BUILD_TARGET)))
   PROJECTS := $(filter-out 03app_dotbot_gateway 03app_dotbot_gateway_lr 03app_sailbot 03app_nrf5340_net,$(PROJECTS))
   ARTIFACT_PROJECTS := 03app_dotbot
+  RADIO_TEST_APPS :=
 endif
 
 # remove incompatible apps (nrf5340, dotbot, gateway) for sailbot-v1 build
 ifeq (sailbot-v1,$(BUILD_TARGET))
   PROJECTS := $(filter-out 01bsp_qdec 01drv_lis3mdl 01drv_move 03app_dotbot_gateway 03app_dotbot_gateway_lr 03app_dotbot 03app_nrf5340_%,$(PROJECTS))
   ARTIFACT_PROJECTS := 03app_sailbot
+  RADIO_TEST_APPS :=
 endif
 
 # remove incompatible apps (nrf5340) for nrf52833dk/nrf52840dk build
@@ -97,6 +100,7 @@ endif
 
 ifneq (,$(filter nrf5340dk-net,$(BUILD_TARGET)))
   ARTIFACT_PROJECTS := 03app_nrf5340_net
+  RADIO_TEST_APPS :=
 endif
 
 SRCS ?= $(shell find bsp/ -name "*.[c|h]") $(shell find crypto/ -name "*.[c|h]") $(shell find drv/ -name "*.[c|h]") $(shell find projects/ -name "*.[c|h]") $(shell find otap/ -name "*.[c|h]") $(shell find radio_test/ -name "*.[c|h]")
@@ -171,4 +175,4 @@ docclean:
        make -C doc/sphinx clean --no-print-directory
 
 doc:
-       make -C doc/sphinx linkcheck html --no-print-directory SPHINXOPTS="-W --keep-going -n"
\ No newline at end of file
+       make -C doc/sphinx linkcheck html --no-print-directory SPHINXOPTS="-W --keep-going -n"
diff --git a/nrf52833dk.emProject b/nrf52833dk.emProject
index 0a165f6..11d5d1c 100644
--- a/nrf52833dk.emProject
+++ b/nrf52833dk.emProject
@@ -64,4 +64,5 @@
   <import file_name="projects/projects-log-dump.emProject" />
   <import file_name="projects/projects-sailbot.emProject" />
   <import file_name="otap/otap.emProject" />
+  <import file_name="radio_test/radio_test.emProject" />
 </solution>
diff --git a/nrf5340dk-app.emProject b/nrf5340dk-app.emProject
index a5ab636..4a9a405 100644
--- a/nrf5340dk-app.emProject
+++ b/nrf5340dk-app.emProject
@@ -65,4 +65,5 @@
   <import file_name="projects/projects-sailbot.emProject" />
   <import file_name="projects/projects-nrf5340-app.emProject" />
   <import file_name="otap/otap.emProject" />
+  <import file_name="radio_test/radio_test.emProject" />
 </solution>
diff --git a/radio_test/radio_test.emProject b/radio_test/radio_test.emProject
index d9651ae..09846d9 100644
--- a/radio_test/radio_test.emProject
+++ b/radio_test/radio_test.emProject
@@ -1,6 +1,6 @@
 <!DOCTYPE CrossStudio_Project_File>
 <solution Name="Radio test" target="8" version="2">
-  <project Name="Rx">
+  <project Name="rx">
     <configuration
       Name="Common"
       project_dependencies="00bsp_dotbot_board(bsp);00bsp_timer(bsp);00bsp_radio(bsp);00bsp_uart(bsp)"
@@ -39,7 +39,7 @@
       <file file_name="$(SeggerThumbStartup)" />
     </folder>
   </project>
-  <project Name="Tx">
+  <project Name="tx">
     <configuration
       Name="Common"
       project_dependencies="00bsp_dotbot_board(bsp);00bsp_timer(bsp);00bsp_radio(bsp)"
@@ -78,7 +78,7 @@
       <file file_name="$(SeggerThumbStartup)" />
     </folder>
   </project>
-  <project Name="Tx idle">
+  <project Name="tx_idle">
     <configuration
       Name="Common"
       project_dependencies="00bsp_dotbot_board(bsp);00bsp_timer(bsp);00bsp_radio(bsp)"
```

File: bsp/nrf/radio.c [hunk 36]
```diff
@@ -255,10 +265,8 @@ void RADIO_IRQHandler(void) {
         NRF_RADIO->EVENTS_DISABLED = 0;
 
         if (radio_vars.state == (RADIO_STATE_BUSY | RADIO_STATE_RX)) {
-            if (NRF_RADIO->CRCSTATUS != RADIO_CRCSTATUS_CRCSTATUS_CRCOk) {
-                puts("Invalid CRC");
-            } else if (radio_vars.callback) {
-                radio_vars.callback(radio_vars.pdu.payload, radio_vars.pdu.length);
+            if (radio_vars.callback) {
+                radio_vars.callback(radio_vars.pdu.payload, radio_vars.pdu.length, NRF_RADIO->CRCSTATUS);
```
Comment (aabadie) [inline bsp/nrf/radio.c:269]:
```suggestion
                radio_vars.callback(radio_vars.pdu.payload, radio_vars.pdu.length, NRF_RADIO->CRCSTATUS != RADIO_CRCSTATUS_CRCSTATUS_CRCOk);
```

File: otap/partition0/main.c [hunk 37]
```diff
@@ -36,9 +36,13 @@ static application_vars_t _app_vars = { 0 };
...
-static void _radio_callback(uint8_t *pkt, uint8_t len) {
-    memcpy(&_app_vars.message_buffer, pkt, len);
-    _app_vars.packet_received = true;
+static void _radio_callback(uint8_t *pkt, uint8_t len, bool crc) {
+    if (!crc) {
+        return;
+    } else {
+        memcpy(&_app_vars.message_buffer, pkt, len);
+        _app_vars.packet_received = true;
+    }
```
Comment (aabadie) [inline otap/partition0/main.c:45]:
```suggestion
    if (!crc) {
        return;
    }
    memcpy(&_app_vars.message_buffer, pkt, len);
    _app_vars.packet_received = true;
```

File: otap/partition1/main.c [hunk 38]
```diff
@@ -36,9 +36,13 @@ static application_vars_t _app_vars = { 0 };
...
-static void _radio_callback(uint8_t *pkt, uint8_t len) {
-    memcpy(&_app_vars.message_buffer, pkt, len);
-    _app_vars.packet_received = true;
+static void _radio_callback(uint8_t *pkt, uint8_t len, bool crc) {
+    if (!crc) {
+        return;
+    } else {
+        memcpy(&_app_vars.message_buffer, pkt, len);
+        _app_vars.packet_received = true;
+    }
```
Comment (aabadie) [inline otap/partition1/main.c:45]:
```suggestion
    if (!crc) {
        return;
    }
    memcpy(&_app_vars.message_buffer, pkt, len);
    _app_vars.packet_received = true;
```

File: projects/03app_dotbot/03app_dotbot.c [hunk 39]
```diff
@@ -96,72 +96,76 @@ static void _update_lh2(void);
...
-    _dotbot_vars.ts_last_packet_received = db_timer_ticks();
-    uint8_t           *ptk_ptr           = pkt;
-    protocol_header_t *header            = (protocol_header_t *)ptk_ptr;
-    // Check destination address matches
-    if (header->dst != DB_BROADCAST_ADDRESS && header->dst != _dotbot_vars.device_id) {
+static void radio_callback(uint8_t *pkt, uint8_t len, bool crc) {
+    if (!crc) {
         return;
-    }
+    } else {
```
Comment (aabadie) [inline projects/03app_dotbot/03app_dotbot.c:102]: Doing if() {} else {} requires to reindent all subsequent code, leading to larger than required changeset. Just add `if () {return;}` and that's it, no need to indent the rest of the function.

File: projects/03app_dotbot_gateway/03app_dotbot_gateway.c [hunk 40]
```diff
@@ -88,13 +88,17 @@ static void _uart_callback(uint8_t data) {
     _gw_vars.uart_queue.last                             = (_gw_vars.uart_queue.last + 1) & (DB_UART_QUEUE_SIZE - 1);
 }
 
-static void _radio_callback(uint8_t *packet, uint8_t length) {
-    if (!_gw_vars.handshake_done) {
+static void _radio_callback(uint8_t *packet, uint8_t length, bool crc) {
+    if (!crc) {
```
Comment (aabadie) [inline projects/03app_dotbot_gateway/03app_dotbot_gateway.c:92]: Same here.

File: projects/03app_sailbot/03app_sailbot.c [hunk 41]
```diff
@@ -182,61 +182,65 @@ int main(void) {
...
-    uint8_t           *ptk_ptr = packet;
-    protocol_header_t *header  = (protocol_header_t *)ptk_ptr;
-
-    // timestamp the arrival of the packet
-    _sailbot_vars.ts_last_packet_received = db_timer_ticks();
-
-    // Check destination address matches
-    if (header->dst != DB_BROADCAST_ADDRESS && header->dst != db_device_id()) {
+void radio_callback(uint8_t *packet, uint8_t length, bool crc) {
+    if (!crc) {
```
Comment (aabadie) [inline projects/03app_sailbot/03app_sailbot.c:186]: And here as well.

File: radio_test/rx/main_radio_test_rx.c [hunk 42]
```diff
@@ -0,0 +1,67 @@
...
+
+static void _radio_callback(uint8_t *packet, uint8_t length, bool crc) {
+    if (!crc) {
+        memcpy(packet + length, &write_crc, 3);
+        db_uart_write(DB_UART_INDEX, packet, length + 3);
+    } else {
+        uint8_t RSSI = db_radio_rssi();
+        memcpy(packet + length, &RSSI, 1);
+        memcpy(packet + length + 1, &uart_end, 1);
+        db_uart_write(DB_UART_INDEX, packet, length + 1);
```
Comment (aabadie) [inline radio_test/rx/main_radio_test_rx.c:45]: This is still not correct, some of my initial comments were lost after renaming the files. Check my first pass of review please.

File: radio_test/tx_idle/main_radio_test_tx_idle.c [hunk 43]
```diff
@@ -0,0 +1,41 @@
...
+
+int main(void) {
+    db_gpio_init(&db_led1, DB_GPIO_OUT);  // Global status
+    db_gpio_set(&db_led1);
+    db_timer_init();
+    db_timer_set_periodic_ms(0, 2000, _led1_blink_slow);
+
+    db_radio_init(&_radio_callback, DB_RADIO_BLE_1MBit);
+    db_radio_set_frequency(8);  // Set the RX frequency to 2408 MHz.
+    db_radio_set_power(RADIO_TXPOWER_TXPOWER_0dBm);
```
Comment (aabadie) [inline radio_test/tx_idle/main_radio_test_tx_idle.c:37]: I would rename this function "db_radio_set_tx_power".

File: bsp/radio.h [hunk 44]
```diff
@@ -67,6 +68,14 @@ void db_radio_set_frequency(uint8_t freq);
  */
 void db_radio_set_channel(uint8_t channel);
 
+/**
+ * @brief Set the power of the radio
+ *
+ *
+ * @param[in] power of the radio
+ */
+void db_radio_set_power(uint8_t power);
```
Comment (aabadie) [inline bsp/radio.h:77]: Rename to `db_radio_set_tx_power`

File: bsp/radio.h [hunk 45]
```diff
@@ -67,6 +68,14 @@ void db_radio_set_frequency(uint8_t freq);
  */
 void db_radio_set_channel(uint8_t channel);
 
+/**
+ * @brief Set the power of the radio
```
Comment (aabadie) [inline bsp/radio.h:72]:
```suggestion
 * @brief Set the TX power of the radio
```

File: bsp/radio.h [hunk 46]
```diff
@@ -67,6 +68,14 @@ void db_radio_set_frequency(uint8_t freq);
...
+ *
+ *
```
Comment (aabadie) [inline bsp/radio.h:74]:
```suggestion
```

File: bsp/radio.h [hunk 47]
```diff
@@ -98,6 +107,15 @@ void db_radio_tx(const uint8_t *packet, uint8_t length);
  */
 void db_radio_rx(void);
 
+/**
+ * @brief Block Radio into TX idle state
+ *
+ * NOTE: Must configure the radio and the frequency before calling this function.
+ * (with the functions db_radio_init db_radio_set_frequency).
```
Comment (aabadie) [inline bsp/radio.h:114]: Not really useful, so can just be removed

File: bsp/radio.h (hunk 46 above)
Comment (aabadie) [reply bsp/radio.h:73]: This is not addressed

File: bsp/radio_ieee_802154.h [hunk 48]
```diff
@@ -0,0 +1,118 @@
+#ifndef __RADIO_H_IEEE_802154
+#define __RADIO_H_IEEE_802154
+
+/**
+ * @defgroup    bsp_radio   Radio support
```
Comment (aabadie) [inline bsp/radio_ieee_802154.h:5]:
```suggestion
 * @defgroup    bsp_radio_ieee_802154   Radio support
```

File: bsp/nrf/radio_ieee_802154.c [hunk 49]
```diff
@@ -0,0 +1,252 @@
+/**
+ * @file
+ * @ingroup bsp_radio
```
Comment (aabadie) [inline bsp/nrf/radio_ieee_802154.c:3]:
```suggestion
 * @ingroup bsp_radio_ieee_802154
```
